### PR TITLE
Add option to obtain velocity directly from hardware

### LINKED ISSUE #159: diff_drive_controller doesn't use joint.velocity
In https://github.com/ros-controls/ros_controllers/blob/indigo-devel/diff_drive_controller/src/odometry.cpp#L78, diff_drive_controller currently differentiates joint.position to obtain velocity, differentiates again to obtain acceleration, and then obtains the final odometry twist component by integrating acceleration.

This results in filtered twist component on the outgoing odometry with a very slow response. Since it's possible to update joint.velocity directly from hardware, would it be possible to generate the twist message directly from this information, improving the response? Could either parametrize or detect from incoming data whether the velocity information is filled out.

### PR DESCRIPTION
**Pull request by paulbovbel**
Fix #159 

Option to use joint velocity from hardware to update odometry twist. Much faster response than rolling mean implementation. Joint position still used in either case to update odometry pose.

### CODE REVIEW

**Comment by efernandez** [comment]
LGTM, but this needs a new test to evaluate the case when `estimate_velocity_from_position: false`.
However, I still wonder if this is the better place to do it, or maybe all this logic should go to the joint level (taking the velocity here, instead of estimating it).

@bmagyar what do you think?

**Comment by paulbovbel** [comment]
If you mean that the robothw implementations should provide the velocity (whether estimated in software or read from hardware), I would agree (proposed this in #159).

**Comment by paulbovbel** [comment]
An alternative could be using NaNs to indicate unused fields in the JointStateHandle, so that diff_drive_controller can decide whether to estimate velocity or not, however this may be getting too clever.

File: diff_drive_controller/src/diff_drive_controller.cpp [hunk 1]
```diff
@@ -178,6 +179,10 @@ namespace diff_drive_controller{
     controller_nh.param("enable_odom_tf", enable_odom_tf_, enable_odom_tf_);
     ROS_INFO_STREAM_NAMED(name_, "Publishing to tf is " << (enable_odom_tf_?"enabled":"disabled"));
 
+    controller_nh.param("estimate_velocity_from_position", estimate_velocity_from_position_, estimate_velocity_from_position_);
+    ROS_INFO_STREAM_NAMED(name_, "Estimating joint velocity from joint position is  " <<
+        (enable_odom_tf_?"enabled":"disabled"));
```
Comment (bmagyar) [inline diff_drive_controller/src/diff_drive_controller.cpp]: This is a bit redundant, I guess you meant `estimate_velocity_from_position_` here :)

**Comment by bmagyar** [comment]
As Enrique mentioned above, we would appreciate a test as well that exercises the param `estimate_velocity_from_position`. Generally I'd add a test for 3 cases and have a very simple test for each and see if the controller works, doesn't work of fails how it is supposed to or not.
- `estimate_velocity_from_position: false`
- `estimate_velocity_from_position: true`
- `estimate_velocity_from_position: something_invalid`

...but in general, it looks good to me.

If you wish we could start a discussion on the velocity estimation topic but I'm afraid that would take more time. I'm more neutral on this one.

**Comment by paulbovbel** [comment]
Typo fixed. I would appreciate any guidance on how you feel I could test this behaviour - `estimate_velocity_from_position` falls back to the default (`true`) if `something_invalid` is entered, and the outputs should be fairly similar in simulation in either the `true` or `false` case.

I'm fairly neutral on further changes to velocity estimation, other than that it should be optional, and it would be great if there was a way for hardware to mark 'unused' fields in JointState handles as `nan` or similar, since 0 is otherwise valid.

**Pull request closed without merging**